Add README and documentation files, re-implement synchronous version of the accountability algorithm based on previous commits, add benchmark scripts to run experiments (#16)
mikanikos/Fork-Accountability: README.md

![](https://github.com/mikanikos/Fork-Accountability/workflows/Build%20and%20Tests/badge.svg)
[![codecov](https://codecov.io/gh/mikanikos/Fork-Accountability/branch/master/graph/badge.svg)](https://codecov.io/gh/mikanikos/Fork-Accountability)

The scope of this project is to implement a simple PoC of a Fork Accountability algorithm in the Tendermint Consensus protocol.

The repository includes simple libraries and modules that allow to run experiments and benchmark the algorithm implementation in a minimal test environment:

- a simple connection library to provide an high-level API for establishing connections and communicating among processes based on a client-server architecture; 

> - an accountability algorithm implementation based on the theoretical specifications given by the research scientists of the DCL lab at EPFL and Interchain Foundation;  

Author: Fixed

- monitor implementation that represents the independent verification entity used to run the accountability algorithm;  

- validator implementation that represents the processes that participate in the Tendermint Consensus protocol; 

- sample test scripts to easily run experiments and test both the accountability algorithm and the interaction between monitor and validators.

## Context and overview
Documentation describing the context and the theoretical background can be found in the [docs](docs) folder.   

## Architecture

> The project is organized in several files and packages in order to guarantee modularity and extendability and maintain at the same time a clear and simple structure.

Author: Fixed

> The monitor (monitor package) is the entity responsible to run the accountability algorithm. It takes as input parameters:

Reviewer: I couldn't find the monitor package, maybe I'm missing something -- did you mean `cmd/monitor/`? Can you add a precise reference to it pls?

Author: Done

- a path of a yaml configuration file which are necessary to initialize and configure the algorithm

- an (optional) path of a file that can be generated to provide detailed information about the whole execution and, especially, of the accountability algorithm

> The monitor is responsible to open the connection with all the validators and initialize the request of the message logs.

Reviewer: With "responsible", it is common to use the phrasing "responsible _for_ X".

```suggestion
The monitor is responsible for opening connections with all the validators and initialize the request of the message logs (described below).
```

Author: Fixed

> Validators are simple processes that listen to a given port and each one has its own messages logs that are the result of the execution of the Tendermint consensus protocol. Message logs and listening port are initialized through a configuration file (different for every validator) along with a unique identifier.

Reviewer: ```suggestion
Validators are simple processes that listen on a given port and each one has its own messages logs that are the result of the execution of the Tendermint consensus protocol. Message logs and listening port are initialized through a configuration file (different for every validator) along with a unique validator identifier.
```

Not sure if we mean _validator_ identifier in the last sentence above, but it would be better to clarify if so.

Author: Yes, validator identifier is more appropriate

Every message in the configuration file must be specified with all the corresponding information associated with it (type, round, height, senderId, possible justifications).

> The monitor will use the connection library to request message logs from all the addresses given in the config file. It will wait for responses from each validator and, as soon as a packet arrives, it will store it and send it to the main thread for running the algorithm if enough messages have been received until that time.

Reviewer: ```suggestion
The monitor will use the connection library to request message logs from all the addresses (i.e., validator processes) given in the config file. It will wait for responses from each validator and, as soon as a packet arrives, it will store it and send it to the main thread. The main thread will run the fork accountability algorithm if enough messages have been received until that time.
```

Author: Fixed

> The request will be repeated after a timeout is expired and if the message received is not valid. If the validator closes a connection or crashes, the monitor will stop waiting for packets from it and will notify the main thread about the failure in the reception.

Reviewer: As a general remark, it is a good idea to avoid passive voice whenever possible. For instance, I suggest a rephrasing below.

```suggestion
The monitor will repeat the request after a timeout expires and if the message received is not valid. If a validator closes a connection or crashes, the monitor will stop waiting for packets from it and will notify the main thread about the failure in the reception.
```

Author: Fixed

The validator, after receiving a valid request packet, will response back if it will have the message logs requested. Otherwise, it will just ignore the request and will not answer the monitor. Optionally, it's possible to configure a response to immediately inform the monitor about the missing log in order to save resources.

The main accountability algorithm is implemented in the accountability package and is described in details in documentation files of the docs folders. Please refer to for a theoretical background or for implementation-specific details.

> The connection library implemented in this project wraps the well-known golang/conn library and provides some abstractions to establish a TCP connection, send and receive TCP packets, serialize and de-serialize messages and listen to a specific port.

Reviewer: For completeness, might be a good idea to add a URL reference to the go pkg here.

Author: Fixed

This library is used by the monitor and the validator to exchange packets for both the request and the sending of the message logs.

## Structure

As an overview, this is the current structure of the project:
    
- [.github](.github): contains Github Actions continuous integration config files

> - [accountability](accountability): contains the main accountability algorithm algorithm

Reviewer: ```suggestion
- [accountability](accountability): contains the main accountability algorithm
```

Author: Fixed

- [cmd](cmd): contains the binaries for the monitor and the validator. Inside each binary folder, there's a folder with sample config files. 

- [common](common): contains abstractions used throughout the project to better handle the input of the algorithm;

- [connection](connection): contains the connection library used by monitor and validators to communicate.

> - [docs](docs): contains some markdown files to describe and documents the protocol and the accountability algorithm from a slightly more theoretical perspective; 

Reviewer: Not sure if my suggestion is good...

```suggestion
- [docs](docs): contains markdown files documenting the project and the accountability algorithm from a slightly more theoretical perspective; 
```

Author: Fixed

                sender: [sender_id]
                round: [round]
                value:
                  data: [value]
                
            
            received_precommit:
              - type: [PREVOTE | PRECOMMIT]
                sender: [sender_id]
                round: [round]
                value:
                  data: [value]

            
            sent_precommit:
              - type: [PREVOTE | PRECOMMIT]
                sender: [sender_id]
                round: [round]
                value:
                  data: [value]

The value in square brackets are values and they are positive integers except for `type` (PREVOTE or PRECOMMIT) and the `data` fields (it can be any integer value, the type can be changed).

The [_config](cmd/validator/_config) folder contains some sample config files for the validator.

### Running test scripts

It's possible to run bash scripts (in a Unix environment) in order to run more validator instances and the monitor at the same time and easily test different scenarios.
A sample bash script is present in the scripts folder and gives a very minimal example of a simple experiment. 

## Acknowledgments

> The project is developed as a semester project in collaboration with the [Distributed Computing Lab](https://dcl.epfl.ch/site/) at [EPFL](https://www.epfl.ch/en/).

Reviewer: You can also add a link to my company, [Informal Systems](informal.systems) if you wish. (We branched out from Interchain Foundation in January.)

Author: Done
